nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/PublishMQTT.java

```diff
@@ -0,0 +1,239 @@
+            logger.warn("Evaluation of the topic property returned null or evaluated to be empty, routing to failure");
+            session.transfer(flowfile, REL_FAILURE);
+            return;
+        }
+
+        // do the read
+        final byte[] messageContent = new byte[(int) flowfile.getSize()];
+        session.read(flowfile, new InputStreamCallback() {
+            @Override
+            public void process(final InputStream in) throws IOException {
+                StreamUtils.fillBuffer(in, messageContent, true);
+            }
+        });
+
+        int qos = context.getProperty(PROP_QOS).evaluateAttributeExpressions(flowfile).asInteger();
+        final MqttMessage mqttMessage = new MqttMessage(messageContent);
+        mqttMessage.setQos(qos);
+        mqttMessage.setPayload(messageContent);
+        mqttMessage.setRetained(context.getProperty(PROP_RETAIN).evaluateAttributeExpressions(flowfile).asBoolean());
+
+        try {
+            /*
+             * Underlying method waits for the message to publish (according to set QoS), so it executes synchronously:
+             *     MqttClient.java:361 aClient.publish(topic, message, null, null).waitForCompletion(getTimeToWait());
+             */
+            mqttClient.publish(topic, mqttMessage);
+            session.transfer(flowfile, REL_SUCCESS);
+        } catch(MqttException me) {
+            logger.error("Failed to publish message.", me);
+            session.transfer(flowfile, REL_FAILURE);
+        }
+    }
+
+    private class PublishMQTTCallback  implements MqttCallback {
+
+        @Override
+        public void connectionLost(Throwable cause) {
+            logger.warn("Connection to " + broker + " lost", cause);
+            try {
+                reconnect();
+            } catch (MqttException e) {
+                logger.error("Connection to " + broker + " lost and re-connect failed");
+            }
+        }
+
+        @Override
+        public void messageArrived(String topic, MqttMessage message) throws Exception {
+            logger.error("Message arrived to a PublishMQTT processor { topic:'" + topic +"; payload:"+ Arrays.toString(message.getPayload())+"}");
+        }
+
+        @Override
+        public void deliveryComplete(IMqttDeliveryToken token) {
+            // Client.publish waits for message to be delivered so this token will always have a null message and is useless in this application.
+            logger.debug("Received 'delivery complete' message from broker for:" + token.toString());
+        }
+    }
+
+
+    private void reconnect() throws MqttException {
+        synchronized (mqttClientConnectLock) {
+            if (!mqttClient.isConnected()) {
+                mqttClient = getMqttClient(broker, clientID, persistence);
+                mqttClient.setCallback(new PublishMQTTCallback());
+                getLogger().info("Connecting to broker: " + broker);
+                mqttClient.connect(connOpts);
+            }
+        }
```

Comment:
About 70-80% of code in this method is the same in both Publish and Consume variants. Perhaps some refactoring into a base method in abstract class would be in order.